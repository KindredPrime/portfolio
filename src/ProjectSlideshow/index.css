.ProjectSlideshow {
  position: relative;
  width: 90%;
  margin-left: auto;
  margin-right: auto;
}

.ProjectSlideshow__screenshot {
  border: 2px solid #A1A1A1;
  width: 100%;
}

.ProjectSlideshow__button {
  position: absolute;
  top: 50%;
  width: 50px;
  height: 50px;
  margin-top: -25px;
  padding: 0px;
  background-color: transparent;
  border: none;
  padding: 0px;
  opacity: 0.5;
}

.ProjectSlideshow__arrow {
  width: 100%;
}

.ProjectSlideshow__button:focus, .ProjectSlideshow__button:hover {
  opacity: 1.0;
}

.ProjectSlideshow__button:hover {
  cursor: pointer;
}

.ProjectSlideshow__button.left {
  left: 10px;
}

.ProjectSlideshow__button.right {
  right: 10px;
}

/*
  Tablet devices
*/
@media screen and (min-width: 690px) {
  .ProjectSlideshow__button {
    width: 75px;
    height: 75px;
    margin-top: -37px;
  }
}

/*
  Desktop/laptop computers
*/
@media screen and (min-width: 1000px) {
  .ProjectSlideshow {
    width: 700px;
  }

  .ProjectSlideshow__button {
    width: 100px;
    height: 100px;
    margin-top: -50px;
  }
}