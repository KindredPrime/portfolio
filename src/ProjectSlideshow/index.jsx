import React from 'react';
import PropTypes from 'prop-types';
import leftArrow from '../images/left-arrow.png';
import rightArrow from '../images/right-arrow.png';
import './index.css';

class ProjectSlideshow extends React.Component {
  state = {
    currentIndex: 0
  };

  nextScreenshot = () => {
    const { currentIndex } = this.state;

    this.setState({
      currentIndex: currentIndex + 1
    });
  }

  prevScreenshot = () => {
    const { currentIndex } = this.state;

    this.setState({
      currentIndex: currentIndex - 1
    });
  }

  render() {
    const { screenshots } = this.props;
    const { currentIndex } = this.state;

    const current = screenshots[currentIndex];

    return (
      <div className="ProjectSlideshow">
        {current &&
          <img
            src={current.src}
            alt={current.alt}
            className="ProjectSlideshow__screenshot"
          />}

        {/*
          Previous screenshot
        */}
        {currentIndex > 0 &&
          <button
            type="button"
            className="ProjectSlideshow__button left"
            onClick={this.prevScreenshot}
          >
            <img
              src={leftArrow}
              alt="previous screenshot"
              className="ProjectSlideshow__arrow"
            />
          </button>}

        {/*
          Next screenshot
        */}
        {currentIndex < screenshots.length - 1 &&
          <button
            type="button"
            className="ProjectSlideshow__button right"
            onClick={this.nextScreenshot}
          >
            <img
              src={rightArrow}
              alt="next screenshot"
              className="ProjectSlideshow__arrow"
            />
          </button>}
      </div>
    );
  }
}

ProjectSlideshow.propTypes = {
  screenshots: PropTypes.arrayOf(
    PropTypes.shape({
      src: PropTypes.string.isRequired,
      alt: PropTypes.string.isRequired
    })
  ).isRequired
}

export default ProjectSlideshow;