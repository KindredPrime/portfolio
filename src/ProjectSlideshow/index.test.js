import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ProjectSlideshow from './index';
import testScreenshot from '../images/test.png';

function renderNextImage(testScreenshots) {
  render(
    <ProjectSlideshow screenshots={testScreenshots} />
  );

  userEvent.click(screen.getByAltText('next screenshot'));
}

describe('ProjectSlideshow Component', () => {
  const testScreenshots = [
    {
      src: testScreenshot,
      alt: 'First test screenshot'
    },
    {
      src: testScreenshot,
      alt: 'Middle test screenshot'
    },
    {
      src: testScreenshot,
      alt: 'Last test screenshot'
    }
  ];

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<ProjectSlideshow screenshots={[]} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the content as expected', () => {
    render(
      <ProjectSlideshow screenshots={testScreenshots} />
    );

    expect(document.body).toMatchSnapshot();
  });

  it('renders both arrow buttons when a middle snapshot is rendered', () => {
    renderNextImage(testScreenshots);

    expect(screen.getByAltText('next screenshot')).toBeInTheDocument();
    expect(screen.getByAltText('previous screenshot')).toBeInTheDocument();
  });

  it(
    'renders only the left arrow when the last of multiple screenshots is rendered',
    () => {
      renderNextImage([testScreenshots[0], testScreenshots[1]]);

      expect(screen.getByAltText('previous screenshot')).toBeInTheDocument();
      expect(screen.queryByAltText('next button')).toEqual(null);
    });

  it(`renders no arrow buttons when there's only one screenshot`, () => {
    render(
      <ProjectSlideshow screenshots={[testScreenshots[0]]} />
    );

    expect(screen.queryByRole('button')).toEqual(null);
  });

  it(
    'renders the next screenshot when the right arrow button is clicked',
    () => {
      renderNextImage(testScreenshots);
      expect(document.body).toMatchSnapshot();
    });

  it(
    'renders the previous screenshot when the left arrow button is clicked',
    () => {
      renderNextImage(testScreenshots);
      userEvent.click(screen.getByAltText('previous screenshot'));
      expect(document.body).toMatchSnapshot();
    });
});