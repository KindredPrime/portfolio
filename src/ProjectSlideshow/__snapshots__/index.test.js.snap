// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectSlideshow Component renders the content as expected 1`] = `
<body>
  <div>
    <div
      class="ProjectSlideshow"
    >
      <img
        alt="First test screenshot"
        class="ProjectSlideshow__screenshot"
        src="test.png"
      />
      <button
        class="ProjectSlideshow__button right"
        type="button"
      >
        <img
          alt="next screenshot"
          class="ProjectSlideshow__arrow"
          src="right-arrow.png"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`ProjectSlideshow Component renders the next screenshot when the right arrow button is clicked 1`] = `
<body>
  <div>
    <div
      class="ProjectSlideshow"
    >
      <img
        alt="Middle test screenshot"
        class="ProjectSlideshow__screenshot"
        src="test.png"
      />
      <button
        class="ProjectSlideshow__button left"
        type="button"
      >
        <img
          alt="previous screenshot"
          class="ProjectSlideshow__arrow"
          src="left-arrow.png"
        />
      </button>
      <button
        class="ProjectSlideshow__button right"
        type="button"
      >
        <img
          alt="next screenshot"
          class="ProjectSlideshow__arrow"
          src="right-arrow.png"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`ProjectSlideshow Component renders the previous screenshot when the left arrow button is clicked 1`] = `
<body>
  <div>
    <div
      class="ProjectSlideshow"
    >
      <img
        alt="First test screenshot"
        class="ProjectSlideshow__screenshot"
        src="test.png"
      />
      <button
        class="ProjectSlideshow__button right"
        type="button"
      >
        <img
          alt="next screenshot"
          class="ProjectSlideshow__arrow"
          src="right-arrow.png"
        />
      </button>
    </div>
  </div>
</body>
`;
